src/server/server_family.cc

```diff
@@ -270,40 +270,113 @@ error_code ServerFamily::LoadRdb(const std::string& rdb_file) {
   return ec;
 }
 
-void AppendMetric(http::StringResponse* resp, absl::AlphaNum name, absl::AlphaNum val) {
-  /**
-   * Gets a metric name and a value and write it using Prometheus format.
-   *
-   * TODO:
-   *  1. add metrics descriptions.
-   *  2. support other types of metrics, not just gauge :)
-   */
-
-  const auto full_name = StrCat("dragonfly_", name);
-  absl::StrAppend(&resp->body(), "# HELP ", full_name, " ", name, "\n");
-  absl::StrAppend(&resp->body(), "# TYPE ", full_name, " gauge\n");
-  absl::StrAppend(&resp->body(), full_name, " ", val, "\n");
+enum MetricType { COUNTER, GAUGE, SUMMARY, HISTOGRAM };
+
+const char* MetricTypeName(MetricType type) {
+  switch (type) {
+    case MetricType::COUNTER:
+      return "counter";
+    case MetricType::GAUGE:
+      return "gauge";
+    case MetricType::SUMMARY:
+      return "summary";
+    case MetricType::HISTOGRAM:
+      return "histogram";
+  }
+  return "unknown";
+};
+
+#define GET_METRIC_FULLNAME(metric_name) StrCat("dragonfly_", metric_name)
+
+void AppendMetricHeader(string* dest, string_view metric_name, string_view metric_help,
+                        MetricType type) {
+  const auto full_metric_name = GET_METRIC_FULLNAME(metric_name);
+  absl::StrAppend(dest, "# HELP ", full_metric_name, " ", metric_help, "\n");
+  absl::StrAppend(dest, "# TYPE ", full_metric_name, " ", MetricTypeName(type), "\n");
+};
+
+void AppendLabelTupple(absl::Span<const string_view> label_names,
+                       absl::Span<const string_view> label_values, string* dest) {
+  if (label_names.empty())
+    return;
+
+  absl::StrAppend(dest, "{");
+  for (size_t i = 0; i < label_names.size(); ++i) {
+    if (i > 0) {
+      absl::StrAppend(dest, ", ");
+    }
+    absl::StrAppend(dest, label_names[i], "=\"", label_values[i], "\"");
+  }
+
+  absl::StrAppend(dest, "}");
+};
+
+void AppendMetricValue(string* dest, string_view metric_name, absl::AlphaNum value,
+                       absl::Span<const string_view> label_names,
+                       absl::Span<const string_view> label_values) {
+  absl::StrAppend(dest, GET_METRIC_FULLNAME(metric_name));
+  AppendLabelTupple(label_names, label_values, dest);
+  absl::StrAppend(dest, " ", value, "\n");
+};
+
+void PrometheusMetricPrinter(http::StringResponse* resp, Metrics m) {
```

Comment:
I prefer naming functions as verbs and use nouns for classes.
For example, `PrintPrometheusMetrics`.

Also, swap the arguments and pass `m` as `const Metrics& m` - it's pretty big struct no need to copy it by value.